core/src/test/scala/unit/kafka/coordinator/transaction/TransactionCoordinatorTest.scala

```diff
@@ -463,31 +463,118 @@ class TransactionCoordinatorTest {
     verify(transactionManager).getTransactionState(ArgumentMatchers.eq(transactionalId))
   }
 
+  @Test
+  def testEndTxnRetryIsAllowedDuringV1UpgradeToV2(): Unit = {
```

Comment:
Sorry I should have caught this sooner, but I'm a bit confused by this case. 
We complete commit with tv 0. 

I don't know if we would retry with tv_2. This would start the next transaction right? And we wouldn't have the previous epoch if we upgraded to tv_2. The only case for a retry for tv0 would be with the same tv and epoch. This test seems to send a request as if the previous request was tv2. Perhaps we should remove this test.

Reply:
You are right. Retry with TV2 is abnormal. The client should use the same TV in one transaction.